src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs

```diff
@@ -108,6 +108,20 @@ public static void BeginInvokeDotNet(JSRuntime jsRuntime, DotNetInvocationInfo i
             // Returned a task - we need to continue that task and then report an exception
             // or return the value.
             task.ContinueWith(t => EndInvokeDotNetAfterTask(t, jsRuntime, invocationInfo), TaskScheduler.Current);
+
+        }
+        else if (syncResult?.GetType().GetGenericTypeDefinition() == typeof(ValueTask<>))
+        {
+            //since syncResult is generic we have to get AsTask method for its type
+            var valueTaskAsTaskMethod = syncResult.GetType().GetMethod("AsTask", BindingFlags.Instance | BindingFlags.Public);
```

Comment:
I think the `MethodInfo` could be cached in a static field.

Reply:
Rather than doing it this way, it's better to call a factory that creates a delegate that converts to this type and cache that delegate.

```
private static Func<Task> CreateConverter(object? result) => ((ValueTask<T>)object).AsTask()
```
And then have a ConcurrentDictionary<Type, Func<Task>> that performs the conversion.

Reply:
Ok so I think I implemented it as You have said but I still had to GetMethod info for this factory method since i had to create generic variants of it runtime. I have made this factory method's MethodInfo cache as static readonly. 
I feel like creating two new fields for this is messy so feel free to suggest something different. I have tried doing it with local function but from what I've read it's not trivial to get MethodInfo on local function